pkg/forecast/spending.go

```diff
@@ -200,37 +193,58 @@ func (s *spendingInstructionBase) getNextSpendingEventAfter(ctx context.Context,
+			// NOTE At the time of writing this, event.RollingAllocation is not being defined anywhere. But this is
+			// ultimately what the math will end up being once it is defined, and we calculate the effects of a transaction.
+			event.RollingAllocation = event.RollingAllocation - s.spending.TargetAmount
+		case nextRecurrence.Equal(fundingFirst.Date):
+			// The next event will be both a contribution and a transaction.
+			event.Date = nextRecurrence
+			event.ContributionAmount = totalContributionAmount
+			event.TransactionAmount = s.spending.TargetAmount
+			event.RollingAllocation = (event.RollingAllocation + totalContributionAmount) - s.spending.TargetAmount
+			event.Funding = fundingFirst.Events
 		}
-		event.RollingAllocation = event.RollingAllocation + totalContributionAmount
-	case nextRecurrence.Before(fundingFirst.Date):
-		// The next event will be a transaction.
-		event.Date = nextRecurrence
-		event.TransactionAmount = s.spending.TargetAmount
-		// NOTE At the time of writing this, event.RollingAllocation is not being defined anywhere. But this is
-		// ultimately what the math will end up being once it is defined, and we calculate the effects of a transaction.
-		event.RollingAllocation = event.RollingAllocation - s.spending.TargetAmount
-	case nextRecurrence.Equal(fundingFirst.Date):
-		// The next event will be both a contribution and a transaction.
-		event.Date = nextRecurrence
-		event.ContributionAmount = totalContributionAmount
-		event.TransactionAmount = s.spending.TargetAmount
-		// NOTE At the time of writing this, event.RollingAllocation is not being defined anywhere. But this is
-		// ultimately what the math will end up being once it is defined, and we calculate the effects of a transaction.
-		event.RollingAllocation = (event.RollingAllocation + totalContributionAmount) - s.spending.TargetAmount
-		event.Funding = []FundingEvent{
-			fundingFirst,
+	case models.SpendingTypeGoal:
+		// Goals are done once there is no more to contribute to them.
+		if totalContributionAmount == 0 {
+			return nil
+		}
+		switch {
+		case fundingFirst.Date.Before(nextRecurrence):
+			// The next event will be a contribution.
+			event.Date = fundingFirst.Date
+			event.ContributionAmount = totalContributionAmount
+			event.Funding = fundingFirst.Events
+			event.RollingAllocation = event.RollingAllocation + totalContributionAmount
+			event.CompletedGoal = event.RollingAllocation >= s.spending.TargetAmount
+		case nextRecurrence.Before(fundingFirst.Date):
+			// No-op, goals do nothing when they are completed.
+			return nil
+		case nextRecurrence.Equal(fundingFirst.Date):
+			// Next event will the completion of the goal.
+			event.Date = nextRecurrence
+			event.ContributionAmount = totalContributionAmount
+			event.RollingAllocation = event.RollingAllocation + totalContributionAmount
+			event.Funding = fundingFirst.Events
+			event.CompletedGoal = event.RollingAllocation >= s.spending.TargetAmount
```

Comment:
Just a thought, should goals and expenses be calculated the same way at all?

Should we have separate spending instructions for each type?